Version Control Systems Strategy v0.01

gabrielcossette opened the pull request:
Closes #1125

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+
+Moving to the digital age requires improving IT's responsiveness.
+
+To improve IT's responsiveness, we must find ways to reduce risks associated with its use.
```
Do we go into what risks come from using multiple VCS, including no VCS at all? Or do we keep that for a fully fleshed out business case?

gabrielcossette replied:
The Benefits section kind of show the "opposite" of the risks ;)
I would keep the details for a fully fleshed out business case.

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+
+### Development and Operations Teams (incl. SSC) <!-- omit in toc -->
+
+1. Use a Git-based Version Control System (VCS) among the following options, in that order of preference:
```
As it's a policy, should there be a pointer to the list of approved VCS instead of listing them here? (just thinking of some laws that still ask for faxes to be used :wink:

gabrielcossette replied:
That's a good point. However, I'm not aware of an existing list with preferences. I assume a good place to create one would be the OSS Mgmt Framework, so the related action "Update ESDC OSS Management Framework" would cover it.

tsanov replied:
IMHO, "1. Use a Git-based Version Control System (VCS) ..." should read "1. Use a DevOps platform ..." - none of the platforms listed below (GitHub, GCCode (a rebranded GitLab), Azure DevOps) is a VCS, most of them are leveraging Git as a VCS but Azure DevOps can also leverage TFS (https://docs.microsoft.com/en-us/azure/devops/user-guide/source-control?view=azure-devops)...

gabrielcossette replied:
Good point, thanks! I specified DevOps platform :)

gcharest replied:
@tsanov I agree with the differentiation between VCS and Coding platform: GitHub, GitLab, etc. are not Git but social coding platform built around Git.

I would also argue that TFVC should be gradually be replaced by Git as the default VCS to ensure the intended outcome: Decentralized VCS with DevOps capabilities and the ability to work in the open, which is part of the rules we have to align with. Not that we are all there yet but it should still be the desired end-state given that this is a strategy.

See this MS article for more details: https://docs.microsoft.com/en-us/azure/devops/repos/git/import-from-tfvc?view=azure-devops

gcharest replied:
Also pointing out here that Azure DevOps now recommends using Git as the default VCS:
https://docs.microsoft.com/en-us/azure/devops/repos/tfvc/comparison-git-tfvc?view=azure-devops

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+    - Wiki (documentation);
+    - Build and Test Automation (e.g., security-related); and
+    - Deployment to internal or external platforms.
+4. Make their projects in the VCS open by default.
```
I'd like to maybe expand on this here so that open by default is properly understood:
- Project space is discoverable
- Teams have established processes to support and encourage contributions from employees outside the team and the organization (public)
- Clear code of conduct is available
- Security Disclosure Process is in place for reporting vulnerabilities and to address them
- Documentation is kept to a minimum but is maintained up to date at all times (anyone should be able to stop by the project and know exactly how to use it.
  - Each release (TBD), documentation is automatically translated (or something like that) 
- Classified information and secrets are kept with secrets management tools/services and never hardcoded in the source code.

Something akin to that.

gcharest replied:
I'd also like to add another rule (or maybe it's outside this strategy...)
- Source code Security and Compliance is automated through the use of key tools and services so that software libraries, dependencies, etc. are assessed every time a new commit to the code base is done.
-  A bill of materials is always automatically generated and made available against the latest commit to the main branch.

gabrielcossette replied:
Awesome points! I added them and also included a link to TBS Guide for Publishing Open Source Code.

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+  <tr>
+    <td rowspan="3"><b>Architecture</b></td>
+    <td>Update ESDC OSS Management Framework</td>
+    <td>Update ESDC OSS Management Framework to reflect the expected usage and capabilities of each approved VCS</td>
```
Should the OSS framework be targeted here? Or should we avoid adding delays for its endorsement?

gabrielcossette replied:
Good point! I've now put the Endorsement before the Update so that it doesn't add delays to the endorsement already in progress.

gabrielcossette replied:
I mean, not done "now" but the commit is upcoming :)

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+    <td><b>DevCoP</b></td>
+    <td>
+      EA<br>
+      ?<br>
```
Senio app devs, Enterprise OPS, BIS, etc.

gcharest replied:
ITSM?

gcharest replied:
APM

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+    <td><b>SSC</b></td>
+    <td>
+      Cloud CoE<br>
+      ?
```
Senior app devs, IT Security, DevOps CoE

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+    </td>
+    <td><b>SSC</b></td>
+    <td>
+    DevCoP<br>
```
DevOps CoE?

gabrielcossette replied:
Do you mean that DevOps CoE is the official name? According to their [website](https://esdc-devcop.github.io), I believe it's DevCoP.

gcharest replied:
The DevOps CoE is not yet created. There's a CoP but we are asking for the creation of the CoE in charge of implementing the TSDM. As such, I'd see them as the leads?

gabrielcossette replied:
Sorry, I forgot about the proposed DevOps CoE, my bad! I'll definitely add them with the mention "(to be created)", thanks!

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+    ?
+    </td>
+  </tr>  
+</table>
```
I'm thinking an action is going to be needed to tackle older languages and projects: how do we bring COBOL and other projects ran in the past related to mainframes for example? I know it's possible, IBM and subject matter experts have been writing about that but that's partly Tech Debt as well: 
- Leverage refactoring of the code to decouple the monolithic systems we had and that we need to adapt to BDM for example?

gabrielcossette replied:
I agree, this is very much needed. However, I think it goes a bit beyond the scope of this Strategy. These projects would still be able to leverage the VCSs in "legacy fashion". The refactoring should be a Tech Debt Strategy on its own in my view.

gcharest commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+</tr>
+</table>
+
+*Manual*: the collection of data requires manual intervention (e.g., surveys, interviews, emails, spreadsheet updates).
```
Regarding metrics, I've been asked to use quantatitive indicators: so increase is not enough, they were asking for how much? If we're currently 20% on Git and rest is distributed in many other, we would like to be at 80% in 2 years.

gabrielcossette replied:
Agreed. This goes into roadmap territories. The Metrics section can simply list them while the Roadmap section would add objectives. I would keep the roadmap for version 0.02

gcharest reviewed (commented):
Hi @gabrielcossette Sorry for the delay, here is my review!

gabrielcossette commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+
+### Business Case
+
+Moving to the digital age requires improving IT's responsiveness.
```
Good points! I tweaked the Benefits section to make sure they are covered. Feel free to propose other edits :)

tsanov commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,346 @@
+: There are 2 types of VCS:
+
+- **Centralized** VCS are based on the idea that there is a single “central” copy of a software project somewhere (most likely on a server), and developers make code changes directly on this central copy.
+- **Decentralized** VCS (DVCS) do not necessarily rely on a central server to store all the versions of a software project's files.
```
I'd suggest "Distributed" instead of "Decentralized", I think this is the standard term.

gabrielcossette replied:
Good catch, thanks! Corrected.

gabrielcossette commented:
Thanks for all your awesome feedback! I believe I addressed all the comments. Let me know if you would like to discuss some more or if you have other comments 😃

remyb2canada commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,366 @@
+
+### Development and Operations Teams (incl. SSC) <!-- omit in toc -->
+
+1. Use a DevOps platform, that includes a Git-based Version Control System (VCS), among the following options, in that order of preference:
```
You will be asked:
- what about mainframes and our COBOL code? Do they have a fit?
- poorly coded code, inherited from decades of technical debt that includes sensitive information (e.g. IP addresses, passwords). We're talking hundres of thousands lines of codes here. I was told by BSI they more than once got ATIPed on their code and had to take efforts in stripping the sensitive part. They have been using certain tools to scan the code of bad practice and static code security vulnerabilities, I'm not sure how far they were able to get. The idea is that it's not as simple as "just clean your code". Just be ready for that.

gabrielcossette replied:
Good thinking, noted 👍

remyb2canada commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,366 @@
+    - Wiki (documentation);
+    - Build and Test Automation (e.g., security-related); and
+    - Deployment to internal or external platforms.
+4. Make their projects in the VCS open by default and adhere to best practices such as:
```
I would split it into two:
4. Make their projects in the VCS open by default (do we mean read-only? Or also contribution as well?)
5. Adhere to best practices such as:
- ...
- ...
- ...

Note that last 5. best practice could also be an action item in the coherent set of actions (Architecture/Define best practices and standards). I.e.
5. Adhere to best practices as listed in the <ESDC best practices and standards> (which is something to write)

gcharest replied:
Working in the open means the team maintains the control on what makes it into the project but supports and encourages contributions... I would propose that this is the default behavior :thinking:

gabrielcossette replied:
Great suggestion! Added to the document. I agree that contributions must be encouraged by default.

remyb2canada commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,366 @@
+  </tr>  
+</table>
+
+## Measuring the Strategy's success
```
Some DORA-related capabilities measurement could be interesting here:
https://www.devops-research.com/research.html#capabilities

gabrielcossette replied:
Good idea! Added DORA metrics :)

gabrielcossette replied:
Good idea! I added DORA VCS metrics for now :)

remyb2canada commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,366 @@
+</div>
+
+<!-- markdownlint-disable MD001 -->
+##### Table of Content <!-- omit in toc -->
```
Recommend adding an Appendix - Traceability Matrix

IITB Way Forward is mentioned but unclear how this strategy supports it.
PwC Independent Study would also be a good Traceability Matrix to add
Information Management Policy (TB Directive on Service and Digital now, but ESDC has an Information Management Strategy too)
GC Architecture Standards
If there's something in the DOSP would be a bonus

gabrielcossette replied:
Very relevant, thanks! Added that new section to the document as well.

remyb2canada reviewed (commented):
Adding recommendations for your brain cells.
J'aime le format!

gcharest commented:
@gabrielcossette I really like this last set of changes. Suggest we publish this first version and start socializing?

gcharest reviewed (approved):
Propose we publish version 0.01 and start socializing w/ tracking feedback.

gcharest commented:
@gabrielcossette Were we ready for iteration 0.1?

gabrielcossette commented on _pages/en/strategy-version-control-systems.md:
```diff
@@ -0,0 +1,366 @@
+The **Guiding Policy**, once operationalized, will define the default systems where ESDC IITB will publish and maintain its source code, scripts, and system/database configurations.
+All ESDC personnel involved in the management of IT Products is expected to adhere to this policy, including but not limited to, development and operations staff.
+
+### Business Case
```
Excellent considerations to keep in mind for the next iteration!
